### PR TITLE
Statistics correction

### PR DESCRIPTION
**Pull request by lapid92**
Add second moment correction in addition to bias correction
  Add second moment correction tests

### CODE REVIEW

File: model_compression_toolkit/core/common/statistics_correction/apply_second_moment_correction_to_graph.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+     Args:
+        graph: Graph to apply second moment correction.
+        representative_data_gen (Callable): Dataset used for calibration.
+        core_config (CoreConfig): Configuration object containing parameters of how the model should be
+         quantized, including mixed precision parameters.
+        fw_info: FrameworkInfo object with information about the specific framework's model.
+        fw_impl: FrameworkImplementation object with a specific framework methods implementation.
+     """
+
+    analyzer_graph(fw_impl.attach_sc_to_node,
```
Comment (elad-c) [inline model_compression_toolkit/core/common/statistics_correction/apply_second_moment_correction_to_graph.py:47]: This code is repeated, consider putting it in a function

File: model_compression_toolkit/core/common/statistics_correction/apply_second_moment_correction_to_graph.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+from model_compression_toolkit.core.common.framework_implementation import FrameworkImplementation
+from model_compression_toolkit.core.common.model_collector import ModelCollector
+from model_compression_toolkit.core.common.quantization.core_config import CoreConfig
+from model_compression_toolkit.core.common.quantization.quantization_analyzer import analyzer_graph
+from model_compression_toolkit.core.common.quantization.quantization_params_generation.qparams_activations_computation import \
+    get_activations_qparams
+from model_compression_toolkit.core.common.substitutions.apply_substitutions import substitute
+
+
+def _assign_act_threshold(graph: Graph,
```
Comment (elad-c) [inline model_compression_toolkit/core/common/statistics_correction/apply_second_moment_correction_to_graph.py:31]: rename to collect and assign...

File: model_compression_toolkit/core/keras/graph_substitutions/substitutions/batchnorm_reconstructing.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+        for qc in bn_node.candidates_quantization_cfg:
+            qc.weights_quantization_cfg.enable_weights_quantization = False
+            qc.activation_quantization_cfg.enable_activation_quantization = False
+
+        graph.add_node_with_in_edges(bn_node, [conv_node])
+        graph.add_edge(bn_node, out_node, **source_edge.get_attributes())
+        graph.remove_edge(conv_node, out_node)
+
+        assert len(graph.nodes) - num_nodes_before_substitution == 1
```
Comment (elad-c) [inline model_compression_toolkit/core/keras/graph_substitutions/substitutions/batchnorm_reconstructing.py:117]: Make this check in tests

File: model_compression_toolkit/core/keras/graph_substitutions/substitutions/batchnorm_refusing.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
+# Copyright 2021 Sony Semiconductor Israel, Inc. All rights reserved.
```
Comment (elad-c) [inline model_compression_toolkit/core/keras/graph_substitutions/substitutions/batchnorm_refusing.py:1]: 2022

File: model_compression_toolkit/core/keras/graph_substitutions/substitutions/batchnorm_refusing.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+def batchnorm_refusing_node_matchers():
+    """
+    Function generates matchers for matching:
+    (DepthwiseConv2D, Conv2D, Conv2DTranspose)[activation=linear] -> BatchNormalization.
+
+    Returns:
+        Matcher for batch norm nodes, and source nodes.
+    """
+    bn_node = NodeOperationMatcher(BatchNormalization)
+    conv_node = NodeOperationMatcher(DepthwiseConv2D) | \
```
Comment (elad-c) [inline model_compression_toolkit/core/keras/graph_substitutions/substitutions/batchnorm_refusing.py:36]: Same question about Dense

File: model_compression_toolkit/core/keras/graph_substitutions/substitutions/batchnorm_refusing.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+                NodeOperationMatcher(Conv2D) | \
+                NodeOperationMatcher(Conv2DTranspose)
+
+    activation_linear = NodeFrameworkAttrMatcher(ACTIVATION, LINEAR)
+    source_node = conv_node & activation_linear
+    return bn_node, source_node
+
+
+def update_kernel_for_bn_refusing_fn(conv_node: BaseNode,
```
Comment (elad-c) [inline model_compression_toolkit/core/keras/graph_substitutions/substitutions/batchnorm_refusing.py:45]: Consider using the function from batch_norm_folding

File: model_compression_toolkit/core/keras/keras_implementation.py [hunk 7]
```diff
@@ -530,3 +571,61 @@ def get_node_mac_operations(self,
...
+        # Move every BN to train mode
+        for layer in model.layers:
+            if isinstance(layer, BatchNormalization):
+                layer.trainable = True
+
+        for _ in tqdm(range(core_config.quantization_config.weights_second_moment_iters)):
+            model(representative_data_gen(), training=True)
+
+        # TODO:
+        # needs to check if gamma&beta are the same?
```
Comment (elad-c) [inline model_compression_toolkit/core/keras/keras_implementation.py:599]: Answer the question and remove the TODO. Probably not, check that in tests

File: model_compression_toolkit/core/keras/keras_implementation.py [hunk 8]
```diff
@@ -530,3 +571,61 @@ def get_node_mac_operations(self,
...
+            representative_data_gen: Dataset to use for retrieving images for the models inputs.
+            graph: Graph to update the parameters after the second moment correction.
+
+        Returns:
+            A function that applies second moment correction.
+        """
+        model = copy.deepcopy(quantized_model)
+
+        # Move every BN to train mode
+        for layer in model.layers:
```
Comment (elad-c) [inline model_compression_toolkit/core/keras/keras_implementation.py:591]: This a potential bug, needto set this for reconstrusted BN only

File: model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/batchnorm_reconstructing.py [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+class BatchNormalizationReconstructing(common.BaseSubstitution):
+    """
+    Reconstruct BatchNormalization after linear layers.
+    """
+    def __init__(self):
+        source_node = NodeOperationMatcher(Conv2d) | \
+                      NodeOperationMatcher(ConvTranspose2d)
+        super().__init__(matcher_instance=source_node)
+
+    def substitute(self,
```
Comment (elad-c) [inline model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/batchnorm_reconstructing.py:38]: Consider moving this function to a parent class in common.

**Review by elad-c — CHANGES_REQUESTED** [review]

File: model_compression_toolkit/core/keras/graph_substitutions/substitutions/batchnorm_refusing.py [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+    Returns:
+        The modified convolution node's weight/kernel/
+    """
+    weights_scale = weights_scale.numpy()
+    if conv_node.type == DepthwiseConv2D:
+        kernel = kernel * weights_scale.reshape(1, 1, kernel.shape[-2], kernel.shape[-1])
+    elif conv_node.type == Conv2DTranspose:
+        kernel = kernel * weights_scale.reshape(1, 1, -1, 1)
+    else:
+        kernel = kernel * weights_scale.reshape(1, 1, 1, -1)
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/keras/graph_substitutions/substitutions/batchnorm_refusing.py:63]: Why not take the channel axis from framework info?

File: model_compression_toolkit/core/keras/graph_substitutions/substitutions/batchnorm_refusing.py [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+                NodeOperationMatcher(Conv2DTranspose)
+
+    activation_linear = NodeFrameworkAttrMatcher(ACTIVATION, LINEAR)
+    source_node = conv_node & activation_linear
+    return bn_node, source_node
+
+
+def update_kernel_for_bn_refusing_fn(conv_node: BaseNode,
+                                     kernel: np.ndarray,
+                                     weights_scale):
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/keras/graph_substitutions/substitutions/batchnorm_refusing.py:47]: Missing type hints

File: model_compression_toolkit/core/keras/graph_substitutions/substitutions/batchnorm_refusing.py [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+from model_compression_toolkit.core.common.graph.graph_matchers import NodeOperationMatcher, \
+    NodeFrameworkAttrMatcher
+from model_compression_toolkit.core.common.substitutions.batchnorm_refusing import BatchNormalizationRefusing
+from model_compression_toolkit.core.keras.constants import KERNEL, LINEAR, ACTIVATION, DEPTHWISE_KERNEL, BIAS, GAMMA, \
+    BETA, \
+    MOVING_MEAN, MOVING_VARIANCE, EPSILON, USE_BIAS, LAYER_NAME
+
+
+def batchnorm_refusing_node_matchers():
+    """
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/keras/graph_substitutions/substitutions/batchnorm_refusing.py:28]: Missing type hints

File: model_compression_toolkit/core/keras/graph_substitutions/substitutions/batchnorm_refusing.py [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+                                     kernel: np.ndarray,
+                                     weights_scale):
+    """
+    Args:
+        conv_node: Convolution node to update the weight/kernel.
+        kernel: The Convolution node's weight
+        weights_scale: Weight scale factor in which to multiply the conv node's weight.
+
+    Returns:
+        The modified convolution node's weight/kernel/
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/keras/graph_substitutions/substitutions/batchnorm_refusing.py:55]: Please fix typo (at the end of the sentence)

File: model_compression_toolkit/core/pytorch/constants.py [hunk 14]
```diff
@@ -48,17 +48,19 @@
 
 # # Layers variables names:
 KERNEL = 'weight'
-KERNEL_SIZE = 'kernel_size'
 BIAS = 'bias'
 GAMMA = 'weight'
 BETA = 'bias'
-IN_CHANNELS = 'in_channels'
 MOVING_MEAN = 'running_mean'
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/pytorch/constants.py:54]: Why did you remove these two constants? (KERNEL_SIZE, IN_CHANNELS)

File: model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/batchnorm_refusing.py [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+                                     kernel: np.ndarray,
+                                     weights_scale: np.ndarray):
+    """
+    Args:
+        conv_node: Convolution node to update the weight/kernel.
+        kernel: The Convolution node's weight
+        weights_scale: Weight scale factor in which to multiply the conv node's weight.
+
+    Returns:
+        The modified convolution node's weight/kernel/
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/batchnorm_refusing.py:49]: Please fix typo, and add missing type hints

File: model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/batchnorm_refusing.py [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+
+from model_compression_toolkit.core.common.graph.graph_matchers import NodeOperationMatcher
+from model_compression_toolkit.core.common import BaseNode
+from model_compression_toolkit.core.common.substitutions.batchnorm_refusing import BatchNormalizationRefusing
+from model_compression_toolkit.core.pytorch.constants import KERNEL, BIAS, GAMMA, BETA, MOVING_MEAN, MOVING_VARIANCE, \
+    EPSILON, USE_BIAS
+
+
+def batchnorm_refusing_node_matchers():
+    """
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/batchnorm_refusing.py:26]: Missing type hints

File: model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/batchnorm_refusing.py [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+    Function generates matchers for matching:
+    (Conv2d, ConvTranspose2d)-> BatchNorm2d.
+
+    Returns:
+        Matcher for batch norm nodes, and source nodes.
+    """
+    bn_node = NodeOperationMatcher(BatchNorm2d)
+    source_node = NodeOperationMatcher(Conv2d) | \
+                  NodeOperationMatcher(ConvTranspose2d)
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/batchnorm_refusing.py:35]: As @elad-c mentioned - shouldn't Linear be here as well?

File: model_compression_toolkit/core/pytorch/pytorch_implementation.py [hunk 18]
```diff
@@ -474,3 +516,66 @@ def get_node_mac_operations(self,
             return kernel_shape[0] * kernel_shape[1]
         else:
             return 0
+
+    def apply_second_moment_correction(self, quantized_model, core_config, representative_data_gen, graph):
+        """
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/pytorch/pytorch_implementation.py:521]: Maybe it's worth considering writing this method in a dedicated file and importing it (Relevant for Keras as well)?

File: model_compression_toolkit/core/pytorch/pytorch_implementation.py [hunk 19]
```diff
@@ -474,3 +516,66 @@ def get_node_mac_operations(self,
             return kernel_shape[0] * kernel_shape[1]
         else:
             return 0
+
+    def apply_second_moment_correction(self, quantized_model, core_config, representative_data_gen, graph):
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/pytorch/pytorch_implementation.py:520]: Missing type hints

File: model_compression_toolkit/core/pytorch/pytorch_implementation.py [hunk 20]
```diff
@@ -474,3 +516,66 @@ def get_node_mac_operations(self,
...
+
+        for _ in tqdm(range(core_config.quantization_config.weights_second_moment_iters)):
+            with torch.no_grad():
+                model(*to_torch_tensor(representative_data_gen()))
+
+        set_model(model)
+
+        # TODO:
+        # needs to check if gamma&beta are the same?
+        # Move every BN to eval mode and update the corresponding BN node params in the graph
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/pytorch/pytorch_implementation.py:549]: Indeed - this should be verified as a test

File: model_compression_toolkit/core/pytorch/pytorch_implementation.py [hunk 21]
```diff
@@ -474,3 +516,66 @@ def get_node_mac_operations(self,
...
+        for name, module in model.named_modules():
+            if isinstance(module, torch.nn.BatchNorm2d):
+                module.eval()
+                bn_node = graph.find_node_by_name(name)[0]
+                bn_node_weights = {GAMMA: module.weight.detach().cpu().numpy(),
+                                   BETA: module.bias.detach().cpu().numpy(),
+                                   MOVING_MEAN: module.running_mean.detach().cpu().numpy(),
+                                   MOVING_VARIANCE: module.running_var.detach().cpu().numpy()}
+                # TODO:
+                # check if to set w by key?
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/pytorch/pytorch_implementation.py:559]: Please remove the TODO

File: model_compression_toolkit/core/pytorch/pytorch_implementation.py [hunk 22]
```diff
@@ -474,3 +516,66 @@ def get_node_mac_operations(self,
...
+                bn_node.weights = copy.deepcopy(bn_node_weights)
+
+    def quantized_model_builder_for_second_moment_correction(self, graph, fw_info):
+        """
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/pytorch/pytorch_implementation.py:563]: This method is duplicated for both Keras and PyTorch implementations. Consider moving it to common

File: model_compression_toolkit/core/common/substitutions/batchnorm_refusing.py [hunk 23]
```diff
@@ -0,0 +1,186 @@
...
+        kernel = conv_node.get_weights_by_keys(self.kernel_str)
+        bias = conv_node.get_weights_by_keys(self.bias_str)
+        gamma = bn_node.get_weights_by_keys(self.gamma_str)
+        beta = bn_node.get_weights_by_keys(self.beta_str)
+        moving_mean = bn_node.get_weights_by_keys(self.moving_mean_str)
+        moving_variance = bn_node.get_weights_by_keys(self.moving__variance_str)
+        eps = bn_node.framework_attr[self.epsilon_str]
+
+        # TODO:
+        # if gamma or beta is None raise Exception
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/common/substitutions/batchnorm_refusing.py:115]: Please remove this TODO

File: model_compression_toolkit/core/common/substitutions/batchnorm_refusing.py [hunk 24]
```diff
@@ -0,0 +1,186 @@
...
+
+        weights_scale = gamma / np.sqrt(moving_variance + eps)
+        bias = beta + (bias - moving_mean) * weights_scale
+
+        kernel, kernel_name = self.update_kernel_for_bn_refusing_fn(conv_node, kernel, weights_scale)
+
+        framework_attr = copy.copy(conv_node.framework_attr)
+        framework_attr[self.use_bias] = True
+        if self.layer_name_str is not None:
+            framework_attr[self.layer_name_str] = conv_node.name + '_refused'
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/common/substitutions/batchnorm_refusing.py:132]: Please use it as a constant (not in constants file, as this is relevant only to this scope, if I'm not missing anything)

File: model_compression_toolkit/core/common/substitutions/batchnorm_refusing.py [hunk 25]
```diff
@@ -0,0 +1,186 @@
...
+
+        weights_dict = {kernel_name: kernel,
+                        self.bias_str: bias}
+
+        conv_bn = copy.deepcopy(conv_node)
+        conv_bn_name = conv_node.name + '_refused'
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/common/substitutions/batchnorm_refusing.py:138]: Same as above-mentioned.

File: model_compression_toolkit/core/common/substitutions/batchnorm_refusing.py [hunk 26]
```diff
@@ -0,0 +1,186 @@
...
+        graph.remove_node(bn_node)
+        graph.remove_node(conv_node)
+
+        assert num_nodes_before_substitution - len(graph.nodes) == 1
+        assert num_edges_before_substitution - len(graph.edges) == 1
+
+        self._calc_weights_quantization_params(conv_bn, weights_scale)
+        return graph
+
+    def _calc_weights_quantization_params(self,
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/common/substitutions/batchnorm_refusing.py:161]: Missing type hints

**Review by reuvenperetz — CHANGES_REQUESTED** [review]

File: model_compression_toolkit/core/runner.py [hunk 27]
```diff
@@ -251,6 +251,11 @@ def get_finalized_graph(initial_graph: Graph,
     if tb_w is not None:
         tb_w.add_graph(transformed_graph, 'after_candidates_filtering')
 
+    ######################################
+    # Statistics Correction
+    ######################################
+    transformed_graph = substitute(transformed_graph, fw_impl.get_substitutions_statistics_correction(quant_config))
```
[the remark this answers is not in the document]
Comment (lapid92) [reply model_compression_toolkit/core/runner.py:257]: The bias correction (compute_bias_correction) should also be called from the ptq_runner?

**Review by elad-c — APPROVED** [review]

**Review by reuvenperetz — APPROVED** [review]

**Pull request merged**